crates/sound/src/device.rs

```diff
@@ -0,0 +1,717 @@
+                        b.stop(stream_id).unwrap();
+                    }
+                }
+                ControlMessageKind::ChmapInfo => {
+                    resp.code = VIRTIO_SND_S_NOT_SUPP.into();
+                }
+            }
+            log::trace!(
+                "returned {} for ctrl msg {:?}",
+                match u32::from(resp.code) {
+                    v if v == VIRTIO_SND_S_OK => "OK",
+                    v if v == VIRTIO_SND_S_BAD_MSG => "BAD_MSG",
+                    v if v == VIRTIO_SND_S_NOT_SUPP => "NOT_SUPP",
+                    v if v == VIRTIO_SND_S_IO_ERR => "IO_ERR",
+                    _ => unreachable!(),
+                },
+                code
+            );
+            desc_chain
+                .memory()
+                .write_obj(resp, desc_hdr.addr())
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+            used_len += desc_hdr.len();
+
+            if vring.add_used(desc_chain.head_index(), used_len).is_err() {
+                log::error!("Couldn't return used descriptors to the ring");
+            }
+            any |= true;
+        }
+
+        // In which cases can happen that we get here and any is false?
+        // PCM_RELEASE and PCM_SET_PARAMS need to be handled asynchronously, therefore
+        // it will be their responsibility to `signal_used_queue`.
+
+        // Send notification if any request was processed
+        if any && vring.signal_used_queue().is_err() {
+            log::error!("Couldn't signal used queue");
+        }
+
+        Ok(!any)
+    }
+
+    fn process_event(&self, _vring: &VringRwLock) -> IoResult<bool> {
+        log::trace!("process_event");
+        Ok(false)
+    }
+
+    fn process_tx(
+        &self,
+        vring: &VringRwLock,
+        audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
+    ) -> IoResult<bool> {
+        let requests: Vec<SoundDescriptorChain> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter(self.mem.as_ref().unwrap().memory())
+            .map_err(|_| Error::DescriptorNotFound)?
+            .collect();
+
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        #[derive(Copy, Clone, PartialEq, Debug)]
+        enum TxState {
+            Ready,
+            WaitingBufferForStreamId(u32),
+            Done,
+        }
+
+        let mut stream_ids = BTreeSet::default();
+
+        for desc_chain in requests {
+            let mut state = TxState::Ready;
+            let mut buffers = vec![];
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            let message = Arc::new(IOMessage {
+                vring: vring.clone(),
+                status: VIRTIO_SND_S_OK.into(),
+                desc_chain: desc_chain.clone(),
+                descriptor: descriptors.last().cloned().unwrap(),
+            });
+            for (_, descriptor) in descriptors.iter().enumerate() {
+                match state {
```

Comment:
Do we have the option of using a for-loop instead? I think it will be easier to read. We know that, in a request, the desc[0] is the header, desc[1..N-2] is the data, and desc[N-1] is the status, or I may be missing some corner-case.

Reply:
This is a for loop. Can you give me a written example of what you mean? In pseudocode etc

Reply:
Yes, it is.  I wonder if we can get rid of the "match state" statement like it is done at https://github.com/virtio-sound/vhost-device/blob/e500ee9576f616e00d38fefde779458b2df8a720/crates/sound/src/vhu_sound.rs#L506 There may have been some corner cases that I missed.

Reply:
Ah, I understand now. Personally the code you linked is really hard to read and understand. Whereas here the type and variant names are making the code more verbose without any performance penalty. Also, illegal states become unrepresentable.